### PR TITLE
W-15943796: Implement an error counting metric.

### PR DESCRIPTION
**Pull request by IvanAndresFritzler**

### CODE REVIEW

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/SpringMuleContextServiceConfigurator.java [hunk 1]
```diff
@@ -247,7 +251,8 @@ public class SpringMuleContextServiceConfigurator extends AbstractSpringMuleCont
       .put(MULE_CORE_EXPORTER_FACTORY_KEY, getBeanDefinition(OpenTelemetrySpanExporterFactory.class))
       .put(MULE_CORE_EVENT_TRACER_KEY, getBeanDefinition(SelectableCoreEventTracer.class))
       .put(MULE_CORE_COMPONENT_TRACER_FACTORY_KEY, getBeanDefinition(CoreEventComponentTracerFactory.class))
-      .put(MULE_METER_PROVIDER_KEY, resolveMeterProvider())
+      .put(MULE_ERROR_METRICS_FACTORY_KEY, resolveErrorMetricsFactory())
+      .put(MULE_CONTAINER_METER_PROVIDER_KEY, resolveContainerMeterProvider())
```
Comment (fsgonz) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/SpringMuleContextServiceConfigurator.java:255]: Is this used or injected in some context? It is placed in the mule context of an artifact.

File: core/src/main/java/org/mule/runtime/core/api/config/MuleProperties.java [hunk 2]
```diff
@@ -233,7 +233,9 @@ public class MuleProperties {
   public static final String MULE_PROFILING_SERVICE_KEY = "_muleProfilingService";
   public static final String MULE_CORE_EVENT_TRACER_KEY = "_muleCoreEventTracer";
   public static final String MULE_CORE_COMPONENT_TRACER_FACTORY_KEY = "_muleCoreComponentTracerFactory";
-  public static final String MULE_METER_PROVIDER_KEY = "_muleMeterProvider";
+  public static final String MULE_ARTIFACT_METER_PROVIDER_KEY = "_muleArtifactMeterProvider";
+  public static final String MULE_CONTAINER_METER_PROVIDER_KEY = "_muleContainerMeterProvider";
```
Comment (fsgonz) [inline core/src/main/java/org/mule/runtime/core/api/config/MuleProperties.java:237]: What is the purpose of the container provider?

File: core/src/main/java/org/mule/runtime/core/internal/context/DefaultMuleContext.java [hunk 3]
```diff
@@ -190,6 +190,7 @@
 import javax.transaction.TransactionManager;
 
 import org.apache.commons.lang3.JavaVersion;
+import org.mule.runtime.metrics.api.internal.error.ErrorMetrics;
```
Comment (fsgonz) [inline core/src/main/java/org/mule/runtime/core/internal/context/DefaultMuleContext.java:193]: unused import

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/SpringMuleContextServiceConfigurator.java [hunk 4]
```diff
@@ -467,12 +480,27 @@ protected MemoryManagementService getMemoryManagementService() {
     return memoryManagementService;
   }
 
-  private static BeanDefinition resolveMeterProvider() {
+  private BeanDefinition resolveArtifactMeterProvider() {
```
Comment (fsgonz) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/SpringMuleContextServiceConfigurator.java:483]: If this is for decorating with the artifactId and supposedly in the context of an artifact you don't want to decorate with the artifactId, I would change the name indicating this is for adding the artifactId as attribute.

File: core/src/main/java/org/mule/runtime/core/privileged/exception/TemplateOnErrorHandler.java [hunk 5]
```diff
@@ -217,13 +228,21 @@ public void dispose() {
       }
 
       @Override
+      // All calling methods will end up transforming any error class other than MessagingException into that one.
       public void accept(Exception error) {
-        // All calling methods will end up transforming any error class other than MessagingException into that one
+        // We must measure the error only once. The following check is to measure it only when entering the error handler logic
+        // and not bubbling inside it.
+        if (!ErrorHandlerContextManager.isHandling((MessagingException) error))
```
Comment (fsgonz) [inline core/src/main/java/org/mule/runtime/core/privileged/exception/TemplateOnErrorHandler.java:235]: I would also add this comment to the jdoc of the isHandling.

**Review by fsgonz — COMMENTED** [review]
As a general strategy, LGTM. ErrorMetrics as an adapter to collaborate with the general Metrics api.

**Review by fsgonz — DISMISSED** [review]

**Review by fsgonz — APPROVED** [review]

File: modules/metrics/mule-metrics-exporter-impl/src/main/java/org/mule/runtime/metrics/exporter/impl/OpenTelemetryMeterExporter.java [hunk 6]
```diff
@@ -114,9 +128,9 @@ public synchronized void registerMeterToExport(org.mule.runtime.metrics.api.mete
 
   @Override
   public void dispose() {
-    counters.forEach(ObservableLongCounter::close);
-    upDownCounters.forEach(ObservableLongUpDownCounter::close);
-
+    longCounters
```
Comment (fsgonz) [inline modules/metrics/mule-metrics-exporter-impl/src/main/java/org/mule/runtime/metrics/exporter/impl/OpenTelemetryMeterExporter.java:131]: ok.

**Comment by IvanAndresFritzler** [comment]
--validate

**Comment by IvanAndresFritzler** [comment]
--validate

**Review by fsgonz — APPROVED** [review]

**Pull request closed without merging**